File: backend/app/schemas/lancamentos.py
"""
Schemas Pydantic — Lançamentos
Alinhados aos endpoints de lançamentos (POST /lancamentos, GET /lancamentos/verificar-pasta).
"""

from datetime import date
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field


class LancamentoCreate(BaseModel):
    """
    Request body para POST /lancamentos.
    RN53 — DC determina qual campo (Deb ou Cred) recebe VValor.
    RN54 — Todos os campos são obrigatórios.
    """

    id_cliente: int = Field(ge=1)
    cod_cliente: int = Field(ge=1)
    dt: date  # Data do lançamento
    conta: int = Field(ge=1)  # "Pasta" na UI — coluna Conta no banco
    ref: str = Field(min_length=1)  # "Histórico" — coluna Ref
    vvalor: Decimal = Field(gt=0)  # Valor do lançamento
    # RN52 — aceita apenas "D" (Débito) ou "C" (Crédito)
    dc: Literal["D", "C"]


class LancamentoCreateResponse(BaseModel):
    """Response do endpoint POST /lancamentos."""

    id: int
    mensagem: str


class VerificarPastaResponse(BaseModel):
    """
    Response do endpoint GET /lancamentos/verificar-pasta.
    RN51 — indica se a pasta já existe para o cliente.
    """

    existe: bool

File: frontend/src/services/lancamentos.ts
/**
 * services/lancamentos.ts
 * Camada de acesso à API de lançamentos.
 * Única camada que conhece as URLs e o contrato HTTP do backend.
 *
 * Equivalente ao fluxo FrmLancaData + FrmLanca (VB.Net).
 * Usa apiFetch (autenticado).
 */

import { apiFetch } from "@/lib/api";

// ---------------------------------------------------------------------------
// Tipos — alinhados aos schemas Pydantic do backend (schemas/lancamentos.py)
// ---------------------------------------------------------------------------

export interface LancamentoCreateRequest {
  id_cliente: number;
  cod_cliente: number;
  dt: string; // "YYYY-MM-DD"
  conta: number; // "Pasta" na UI
  ref: string; // "Histórico"
  vvalor: string; // Decimal como string para evitar ponto flutuante
  dc: "D" | "C";
}

export interface LancamentoCreateResponse {
  id: number;
  mensagem: string;
}

export interface VerificarPastaResponse {
  existe: boolean;
}

// ---------------------------------------------------------------------------
// Funções de serviço
// ---------------------------------------------------------------------------

/**
 * Verifica se a pasta (Conta) já existe para o cliente informado.
 * RN51 — equivalente à query ADODB em txtPasta_Validated no FrmLanca.vb.
 */
export async function verificarPasta(
  idCliente: number,
  pasta: number,
): Promise<VerificarPastaResponse> {
  const params = new URLSearchParams({
    id_cliente: String(idCliente),
    pasta: String(pasta),
  });

  const res = await apiFetch(`/lancamentos/verificar-pasta?${params}`);

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? "Erro ao verificar pasta");
  }

  return res.json();
}

/**
 * Registra um novo lançamento de débito ou crédito.
 * RN52 — DC aceita apenas 'D' ou 'C'.
 * RN53 — DC determina qual campo (Deb ou Cred) recebe VValor (lógica no backend).
 * RN54 — todos os campos são obrigatórios.
 * Equivalente a btnRibSalvar_Click + TableAdapterManager.UpdateAll() no FrmLanca.vb.
 */
export async function registrarLancamento(
  dados: LancamentoCreateRequest,
): Promise<LancamentoCreateResponse> {
  const res = await apiFetch("/lancamentos/", {
    method: "POST",
    body: JSON.stringify(dados),
  });

  if (!res.ok) {
    const err = await res.json().catch(() => ({}));
    throw new Error(err.detail ?? "Erro ao registrar lançamento");
  }

  return res.json();
}
